atom/browser/api/atom_api_web_contents.cc

```diff
@@ -898,8 +899,27 @@ void WebContents::DidStartNavigation(
   }
   bool is_same_document = navigation_handle->IsSameDocument();
   auto url = navigation_handle->GetURL();
-  Emit("did-start-navigation", url, is_same_document, is_main_frame,
-       frame_process_id, frame_routing_id);
+  return Emit(event, url, is_same_document, is_main_frame, frame_process_id,
+              frame_routing_id);
+}
+
+void WebContents::DidStartNavigation(
+    content::NavigationHandle* navigation_handle) {
+  EmitNavigationEvent("did-start-navigation", navigation_handle);
+}
+
+void OnStopSoon(WebContents* web_contents) {
+  if (web_contents && !web_contents->IsDestroyed())
+    web_contents->Stop();
+}
+
+void WebContents::DidRedirectNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (EmitNavigationEvent("will-redirect", navigation_handle)) {
+    scoped_refptr<base::SingleThreadTaskRunner> task_runner(
+        base::ThreadTaskRunnerHandle::Get());
+    task_runner->PostTask(FROM_HERE, base::BindOnce(&OnStopSoon, this));
```

Comment:
Does the asynchronous nature of this leave a window of opportunity for the redirect to start happening? It'd be awesome if there was a way to guarantee that not a single byte of network traffic leaves to the domain of the redirect's target.

Reply:
> It'd be awesome if there was a way to guarantee that not a single byte of network traffic leaves to the domain of the redirect's target.

Its possible by registering navigation throttlers https://cs.chromium.org/chromium/src/content/public/browser/navigation_throttle.h?l=146